---
title: "Multiomic data integration with MOFA2 model"
output: html_notebook
---

```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
library(MOFA2)
library(reticulate)
library(DESeq2)
library(sva)
library(MultiAssayExperiment)
library(tidyverse)
library(BiocParallel)
```

PREPARE MOFA2 MODEL

```{r}
#Load datasets
load("CLLdata/mofaCLL-main/data/gene.Rdata")
load("CLLdata/mofaCLL-main/data/rna.Rdata")
load("CLLdata/mofaCLL-main/data/meth.Rdata")
load("CLLdata/exprMat2.Rdata")
load("CLLdata/mofaCLL-main/data/drug.Rdata")
```

Prepare each single omic dataset

```{r}
#RNA sequencing
exprMat4 <- as.matrix(exprMat4)
exprMat4 <- assay(exprMat4)

##DNA methylation array
methData <- assays(meth)[["beta"]]
nTop = 5000
sds <- genefilter::rowSds(methData)
methData <- methData[order(sds, decreasing = T)[1:nTop],]

##Genetics
gene <- t(gene)

#Drug screening
viabMat <- mutate(drug, id = paste0(Drug,"_",concIndex)) %>%
  select(patientID, id, normVal) %>%
  spread(key = patientID, value = normVal) %>%
  data.frame() %>% column_to_rownames("id") %>% 
  as.matrix()
```

Assemble multiAssayExperiment object

```{r}
#Create object
mofaData <- list(mRNA = exprMat4,
                 Mutations = gene, 
                 Methylation = methData,
                 Drugs = viabMat)

# Create MultiAssayExperiment object 
mofaData <- MultiAssayExperiment::MultiAssayExperiment(
  experiments = mofaData
)

#Only keep samples that have at least three assays
useSamples <- MultiAssayExperiment::sampleMap(mofaData) %>%
  as_tibble() %>% group_by(primary) %>% summarise(n= length(assay)) %>%
  filter(n >= 3) %>% pull(primary)
mofaData <- mofaData[,useSamples]

```

```{r}
#Dimensions for each dataset
experiments(mofaData)
```

```{r}
#How many samples have the complete datasets
table(table(sampleMap(mofaData)$primary))
```

Build MOFA object from multiAssayExperiment object

```{r}
MOFAobject <- create_mofa(mofaData)
MOFAobject
```

```{r}
plot_data_overview(MOFAobject)
```

Setup MOFA training parameters

```{r}
#Define data options
data_opts <- get_default_data_options(MOFAobject)
data_opts
```

Define model options

```{r}
model_opts <- get_default_model_options(MOFAobject)
model_opts$num_factors <- 30
model_opts
```

Define training options

```{r}
train_opts <- get_default_training_options(MOFAobject)
train_opts$convergence_mode <- "slow"
train_opts$seed <- 42
train_opts$maxiter <- 5000
train_opts$drop_factor_threshold <- 0.02
train_opts
```

Run MOFA model

```{r}
MOFAobject <- prepare_mofa(
  object = MOFAobject,
  data_options = data_opts,
  model_options = model_opts,
  training_options = train_opts
)
MOFAobject <- run_mofa(MOFAobject, outfile = "MOFA_CLL.Rdata")
```

Part 2. Overview of MOFA output

```{r}
load("CLLdata/mofaCLL-main/data/gene.Rdata")
load("CLLdata/mofaCLL-main/data/survival.Rdata")
load("CLLdata/reactomeGS2.Rdata")
load("CLLdata/mofaCLL-main/data/doublingTime.Rdata")
load("CLLdata/mofaCLL-main/data/demographic.Rdata")

```

```{r}
library(MOFA2)
library(tidyverse)
library(mltools)
library(survival)
library(survminer)
library(maxstat)
library(gridExtra)
library(car)
library(cowplot)
library(egg)
```

```{r}
plot_factor_cor(MOFAobject)
```

Variance explained by MOFA2 for each omic data

```{r}
# Calculate the variance explained (R2) per factor in each view 
r2 <-calculate_variance_explained(MOFAobject)
r2$r2_total
```

```{r}
r2$r2_per_factor
```

```{r}
#Total variance explained per view
plot_variance_explained(MOFAobject, plot_total = T)[[2]]
```

Variance explained per factor

```{r}
varExpPlot <- plot_variance_explained(MOFAobject, censor = 0.25)
varExpPlot
```

Get weights and factors

```{r}
allWeights <- get_weights(MOFAobject,
                         views = "all",
                         factors = "all",
                         as.data.frame = TRUE) %>% as_tibble() %>%
  mutate(feature = ifelse(feature == "IGHV.status","IGHV",feature),
         factor = gsub("LF","F",factor))

allFactors <- get_factors(
  MOFAobject, 
  factors = "all",
  as.data.frame = TRUE
) %>% as_tibble() %>%
  mutate(factor = gsub("LF","F", factor))

patAnno <- gene[c("IGHV", "trisomy12")] %>% data.frame() %>%
  rownames_to_column("sample") %>%
  filter(!is.na(IGHV),!is.na(trisomy12)) %>%
  mutate(IGHV = ifelse(IGHV==1, "M","U"),
         trisomy12 = ifelse(trisomy12 ==1, "yes","no"))

allFactors <- left_join(allFactors, 
                        patAnno,
                        by = "sample")

```

Factor 1 Profile

```{r}
plot_factor(MOFAobject, 
            factors = 1, 
            color_by = "Factor1"
)
```

Feature Weights

```{r}
#Factor1 Somatic Mutations
plot_weights(MOFAobject,
             view = "Mutations",
             factor = 1,
             nfeatures = 10,     # Top number of features to highlight
             scale = T           # Scale weights from -1 to 1
)
```

Loadings of genomic variations

```{r}
plot_top_weights(MOFAobject,
                 view = "Mutations",
                 factor = 1,
                 nfeatures = 10,     # Top number of features to highlight
                 scale = T           # Scale weights from -1 to 1
)
```

Factor1 mRNA Expression

```{r}
plot_weights(MOFAobject,
             view = "mRNA",
             factor = 1,
             nfeatures = 10,     
             scale = T           
)

```

```{r}
#Heatap for the clusters of Factor1 gene expression
plot_data_heatmap(MOFAobject, 
                  view = "mRNA",
                  factor = 1,  
                  features = 25,
                  denoise = TRUE,
                  cluster_rows = FALSE, cluster_cols = FALSE,
                  show_rownames = TRUE, show_colnames = FALSE,
                  scale = "row"
)
```

Associations between Factor 1 and epigenetic subtypes

```{r}
#Methylation profile
plotTab <- filter(allFactors, factor %in% c("Factor1", "Factor2")) %>%
  mutate(Epitype = methCluster[sample]) %>%
  spread(key = factor ,value = value)

violinLF1_Meth <- ggplot(filter(plotTab, !is.na(Epitype)), aes(x=Epitype, y=Factor1, fill = Epitype)) +
  geom_violin() + geom_point() + scale_fill_manual(values = colList[4:length(colList)]) +
  theme_full + theme(legend.position = "none", axis.title.y = element_text(vjust=-2)) +
  xlab("") + ggtitle("Epigenetic subtypes")
violinLF1_Meth
```

Factor2 profile

```{r}
plot_factor(MOFAobject, 
            factors = 2, 
            color_by = "Factor2"
)
```

```{r}
plot_top_weights(MOFAobject,
                 view = "Mutations",
                 factor = 2,
                 nfeatures = 10,     # Top number of features to highlight
                 scale = T )
```

Plot the separation of the samples by the first two factors

```{r}
plotTab <- filter(allFactors, factor %in% c("Factor1","Factor2")) %>%
  spread(key = factor, value = value) %>% mutate(trisomy12 = factor(trisomy12)) %>%
  filter(!is.na(IGHV), !is.na(trisomy12))

pcaLF1 <- ggplot(plotTab, aes(x=Factor1, y=Factor2, color = trisomy12, 
                              shape = IGHV, label = sample)) + 
  geom_point(size=3) +
  scale_shape_manual(values = c(M = 16, U =1)) +
  scale_color_manual(values = c(no = colList[1], yes = colList[2])) +
  theme_full
pcaLF1
```

Factor3 Profile

```{r}
plot_data_heatmap(MOFAobject, 
                  view = "mRNA",
                  features = 30,
                  factor = 3,  
                  denoise = TRUE,
                  cluster_rows = FALSE, cluster_cols = FALSE,
                  show_rownames = TRUE, show_colnames = FALSE,
                  scale = "row"
)
```

Factor 4 profile

```{r}
plot_data_heatmap(MOFAobject, 
                  view = "mRNA",
                  features = 30,
                  factor = 4, 
                  denoise = TRUE,
                  cluster_rows = FALSE, cluster_cols = FALSE,
                  show_rownames = TRUE, show_colnames = FALSE,
                  scale = "row"
)
```
Factor 5 profile

```{r}
plot_data_heatmap(MOFAobject, 
                  view = "mRNA",
                  features = 10,
                  factor = 5, 
                  denoise = TRUE,
                  cluster_rows = FALSE, cluster_cols = TRUE,
                  show_rownames = TRUE, show_colnames = FALSE,
                  scale = "row"
)
```

Assocations with RNAseq batch

```{r}
allFactors <- get_factors(
  MOFAobject, 
  factors = "all",
  as.data.frame = TRUE
) %>% as_tibble() %>%
  mutate(factor = gsub("LF","F", factor))

batchTab <- filter(allFactors , factor == "Factor5") %>%
  mutate(batch = rna[,match(sample,colnames(rna))]$batch) %>% 
  filter(!is.na(batch)) %>%
  mutate(batch = paste0("batch", batch+1))
```

```{r}
pval <- car::Anova(lm(value ~ factor(batch), batchTab))$`Pr(>F)`[1]
pval <- formatNum(pval, digits = 2)
pAnno <- bquote(italic("P")~"="~.(pval))
colListNew <- colList[-4]
pL4 <- ggplot(batchTab, aes(x=batch, y = value, col = batch)) +
  geom_boxplot() + 
  ggbeeswarm::geom_beeswarm() + 
  scale_color_manual(values = colListNew) +
  annotate("text", x=Inf, y=Inf, label=pAnno, hjust=1.5, vjust=1.5)+
  theme_full +
  theme(legend.position = "none") +
  ylab("F5 value") + xlab("") + ggtitle("F5 ~ RNAseq batch")
pL4
```

Factor 6 profile

```{r}
plot_top_weights(MOFAobject,
                 view = "mRNA",
                 factor = 6,
                 nfeatures = 15, #Top number of features to highlight
                 scale = T           # Scale weights from -1 to 1
)
```

ASSOCIATIONS OF LATENT FACTORS TO CLINICAL BEHAVIORS

Univariate cox regression

```{r}
testTab <- left_join(allFactors, survival, by = c(sample = "patientID"))

#for OS
resOS <- filter(testTab, !is.na(OS)) %>%
  group_by(factor) %>%
  do(comSurv(.$value, .$OS, .$died, TRUE)) %>% ungroup() %>%
  arrange(p) %>% mutate(p.adj = p.adjust(p, method = "bonferroni")) %>%
  mutate(Endpoint = "OS")

#for TTT
resTTT <- filter(testTab, !is.na(TTT)) %>%
  group_by(factor) %>%
  do(comSurv(.$value, .$TTT, .$treatedAfter, TRUE)) %>% ungroup() %>%
  arrange(p) %>% mutate(p.adj = p.adjust(p, method = "bonferroni")) %>%
  mutate(Endpoint = "TTT")
```

Overall survival (OS)

```{r}
resOS
```

Time to treatment (TTT)

```{r}
resTTT
```

Plot p values and hazard ratios
```{r}
plotTab <- bind_rows(resOS, resTTT) %>%
  filter(factor %in% c("Factor1","Factor2","Factor3", "Factor4", "Factor5", "Factor6", "Factor7", "Factor8"))

haPlot <- ggplot(plotTab, aes(x=factor, y = HR, col = Endpoint, dodge = Endpoint)) + 
  geom_hline(yintercept = 1, linetype = "dotted") +
  geom_point(position = position_dodge(width=0.8)) +
  geom_errorbar(position = position_dodge(width =0.8), 
                aes(ymin = lower, ymax = higher), width = 0.3, size=1) + 
  geom_text(position = position_dodge2(width = 0.8),
            aes(x=as.numeric(as.factor(factor))+0.15,
                label = sprintf("italic(P)~'='~'%s'",
                                formatNum(p))),
            color = "black",size =4, parse =TRUE) +
  xlab("Factor") + ylab("Hazard ratio") +
  scale_y_log10(limits = c(0.5,4)) +
  coord_flip() + theme_full + theme(legend.title = element_blank(),
                                    legend.position = c(0.2,0.1),
                                    legend.background = element_blank(),
                                    legend.key.size = unit(0.5,"cm"),
                                    legend.key.width = unit(0.6,"cm"),
                                    legend.text = element_text(size=rel(1.2))) +
  scale_color_manual(values = c(OS = colList[3], TTT = colList[5], TFT = colList[4])) 

haPlot
```
Kaplan-Meiler plots
```{r}
#KM plot for overall survival (OS)
facList <- sort(filter(resOS, p<=0.05)$factor)
osList <- lapply(facList, function(x) {
  eachTab <- filter(testTab, factor == x) %>%
    select(value, OS, died) %>% filter(!is.na(OS))
  pval <- filter(resOS, factor == x)$p
  km(eachTab$value, eachTab$OS, eachTab$died, sprintf("%s VS OS time", x),
     stat = "maxstat", pval = pval, showTable = TRUE)
})

grid.arrange(grobs = osList, ncol = 2)

```
KM plot for time to treatment (TTT)
```{r}
facList <- sort(filter(resTTT, p<=0.01)$factor)
tttList <- lapply(facList, function(x) {
  eachTab <- filter(testTab, factor == x) %>%
    select(value, TTT, treatedAfter) %>% filter(!is.na(TTT))
  pval <- filter(resTTT, factor == x)$p
  km(eachTab$value, eachTab$TTT, eachTab$treatedAfter, sprintf("%s VS TTT", x), stat = "maxstat",
     maxTime = 7, pval = pval, showTable = TRUE)
})

tttList
```
KM plot for subgroup defined by IGHV status and median latent factor values

```{r}
groupTab <- filter(testTab, factor == "Factor4", !is.na(value), !is.na(IGHV)) %>%
  mutate(subgroup = ifelse(value > median(value), paste0(IGHV,"_highFactor4"), paste0(IGHV,"_lowFactor4"))) 

plotList <- list()

# TTT
plotList[["TTT"]] <- km(groupTab$subgroup, groupTab$TTT, groupTab$treatedAfter, "TTT", stat = "binary", maxTime = 7, showP = TRUE, showTable = TRUE, yLabelAdjust = -10)

#OS
# OS
plotList[["OS"]] <- km(groupTab$subgroup, groupTab$OS, groupTab$died, "OS", stat = "binary", maxTime = 7, showP = TRUE, showTable = TRUE, yLabelAdjust = -10)

grid.arrange(grobs = plotList, ncol = 2)
```

Multi-variate Cox regression for Factor 4
```{r}
#Prepare data for multi-variate Cox regression
survTab <- survival
facTab <- filter(allFactors, factor == "Factor4")
riskTab <- gene[,c("IGHV","TP53","NOTCH1","del17p","SF3B1")] %>% 
  data.frame() %>% rownames_to_column("patientID") %>%
  mutate(`TP53.del17p` = as.numeric(TP53 | del17p)) %>%
  select(-TP53, -del17p) %>%
  mutate_if(is.numeric, as.factor) %>%
  left_join(select(demographic, patientID, age, sex), by = "patientID") %>%
  mutate(age = age/10) %>%
  mutate(F4 = facTab[match(patientID, facTab$sample),]$value,
         IGHV = factor(IGHV, levels = c(0,1))) %>%
  dplyr::rename(IGHV_mutated = IGHV, Sex_male = sex, Age = age) %>%
  filter(!is.na(F4))
```

Time to treatment
```{r}
resTTT <- runCox(survTab, riskTab, "TTT", "treatedAfter")
#summary(surv1)
haTTT <- plotHazard(resTTT, title = "TTT") +
  scale_y_log10(limits=c(0.2,5))
haTTT
```

Overall survival
```{r}
resOS <- runCox(survTab, riskTab, "OS", "died")
#summary(surv1)
haOS <- plotHazard(resOS, title = "OS") +
  scale_y_log10(limits=c(0.2,5))
haOS
```

Correlation between F6  and demographics
```{r}
#Pretreatment
plotTab <- left_join(demographic, facTab, by = c(patientID = "sample")) %>%
  filter(!is.na(value)) %>%
  mutate(pretreat = ifelse(is.na(pretreat),NA,
                           ifelse(pretreat ==1 , "yes","no")),
         sex = ifelse(is.na(sex),NA,
                           ifelse(sex =="f" , "Female","Male")))

corRes <- t.test(value ~ pretreat, plotTab)
annoP <- paste("italic(P)~'='~",formatNum(corRes$p.value, digits = 1, format = "e"))

plotTab <- filter(plotTab, !is.na(pretreat)) %>%
  group_by(pretreat) %>% mutate(n=n()) %>% ungroup() %>%
  mutate(pretreat = sprintf("%s\n(n=%s)",pretreat,n))

plotTreat <- ggplot(filter(plotTab,!is.na(pretreat)), aes(x = pretreat, y = value)) + 
  geom_violin(aes(fill = pretreat)) +
  geom_point() + 
  annotate("text", x = -Inf, y = Inf, label = annoP,
           hjust=-0.2, vjust =2, size = 5, parse = TRUE, col= colList[1]) +
  scale_fill_manual(values = colList[3:length(colList)]) +
  ylab("F6") + xlab("Pretreatment") +
  theme_full + theme(legend.position = "none")
plotTreat

```

Association between F4 and outcomes in treatment-naive patients

Univariate Cox regression
```{r}
survival.untreated <- survival %>% mutate(pretreat = demographic[match(patientID, demographic$patientID),]$pretreat) %>%
  filter(pretreat ==0)
testTab <- left_join(allFactors, survival.untreated, by = c(sample = "patientID"))

#for OS
resOS <- filter(testTab, !is.na(OS), factor == "Factor4") %>%
  group_by(factor) %>%
  do(comSurv(.$value, .$OS, .$died, TRUE)) %>% ungroup() %>%
  arrange(p) %>% mutate(p.adj = p.adjust(p, method = "BH")) %>%
  mutate(Endpoint = "OS")

resOS

```


```{r}
#for TTT
resTTT <- filter(testTab, !is.na(TTT), factor == "Factor4") %>%
  group_by(factor) %>%
  do(comSurv(.$value, .$TTT, .$treatedAfter, TRUE)) %>% ungroup() %>%
  arrange(p) %>% mutate(p.adj = p.adjust(p, method = "BH")) %>%
  mutate(Endpoint = "TTT")
resTTT
```

Kaplan-Meiler plots

```{r}
#KM plot for overall survival (OS)
eachTab <- filter(testTab, factor == "Factor4") %>%
  select(value, OS, died) %>% filter(!is.na(OS))
pval <- filter(resOS, factor == "Factor4")$p
kmOS.untreat <- km(eachTab$value, eachTab$OS, eachTab$died, "F4 VS OS time",
                   stat = "maxstat", pval = pval, showTable = TRUE)

kmOS.untreat
```

KM plot for time to treatment (TTT)
```{r}
eachTab <- filter(testTab, factor == "Factor4") %>%
  select(value, TTT, treatedAfter) %>% filter(!is.na(TTT))
pval <- filter(resTTT, factor == "Factor4")$p
kmTTT.untreat <- km(eachTab$value, eachTab$TTT, eachTab$treatedAfter, "F4 VS TTT",
                    stat = "maxstat", pval = pval, showTable = TRUE)

kmTTT.untreat
```

Multi-variate Cox regression
```{r}
#Prepare data for multi-variate Cox regression
survTab <- survival.untreated
facTab <- filter(allFactors, factor == "Factor4")
riskTab <- gene[,c("IGHV","TP53","del17p","SF3B1")] %>% 
  data.frame() %>% rownames_to_column("patientID") %>%
  mutate(`TP53.del17p` = as.numeric(TP53 | del17p)) %>%
  select(-TP53, -del17p) %>%
  mutate_if(is.numeric, as.factor) %>%
  left_join(select(demographic, patientID, age, sex), by = "patientID") %>%
  mutate(age = age/10) %>%
  mutate(Factor4 = facTab[match(patientID, facTab$sample),]$value,
         IGHV = factor(IGHV, levels = c(0,1))) %>%
  dplyr::rename(IGHV_mutated = IGHV, Sex_male = sex, Age = age) %>%
  filter(!is.na(Factor4))
```

Time to treatment

```{r}
resTTT <- runCox(survTab, riskTab, "TTT", "treatedAfter")

#summary(surv1)
haTTT.untreat <- plotHazard(resTTT, title = "Time to treatment")
haTTT.untreat
```

Overall survival
```{r}
resOS <- runCox(survTab, riskTab, "OS", "died")

#summary(surv1)
haOS.untreat <- plotHazard(resOS,"Overall survival")
haOS.untreat
```
Correlation between F4 and and Lymphocyte doubling time

Univariate test
```{r}
#Pearson’s correlation test
LDT <- doublingTime %>% mutate(Factor4 = facTab[match(patID, facTab$sample),]$value) %>%
  filter(!is.na(Factor4)) %>%
  mutate(IGHV = as.factor(facTab[match(patID, facTab$sample),]$IGHV))
corRes <- cor.test(log10(LDT$doubling.time), LDT$Factor4)

```

Scatter plot of correlations
```{r}
pval <- formatNum(corRes$p.value, digits = 1, format = "e")
annoN <- sprintf("n = %s", nrow(LDT))
annoP <- bquote(italic("P")~"="~.(pval))
annoCoef <- sprintf("coefficient = %1.2f",corRes$estimate)

corPlot <- ggplot(LDT, aes(x = Factor4, y = doubling.time/30)) + 
  geom_point(fill =colList[5], shape =21, size=3) + 
  geom_smooth(method = "lm", se=FALSE, color = "grey50", linetype ="dashed" ) +
  annotate("text", x = max(LDT$Factor4), y = Inf, label = annoN,
           hjust=1, vjust =1.5, size = 5, parse = FALSE, col= colList[1]) +
  annotate("text", x = max(LDT$Factor4), y = Inf, label = annoP,
           hjust=1, vjust =3.5, size = 5, parse = FALSE, col= colList[1]) +
  annotate("text", x = max(LDT$Factor4), y = Inf, label = annoCoef,
           hjust=1, vjust =5.5, size = 5, parse = FALSE, col= colList[1]) +
  ylab(bquote("doubling time (months)")) + ggtitle("Lymphocyte doubling time") +
  scale_y_log10() +
  theme_full

corPlot
```

Consider IGHV status
```{r}
#IGHV as a covariate
#ANOVA test
LDT <- doublingTime %>% mutate(Factor4 = facTab[match(patID, facTab$sample),]$value,
                               IGHV = as.factor(facTab[match(patID, facTab$sample),]$IGHV)) %>%
  filter(!is.na(Factor4),!is.na(IGHV))

corRes <- car::Anova(lm(log10(doubling.time) ~ Factor4 + IGHV, data = LDT))
corRes
```

Only M-CLL
```{r}
LDT.M <- filter(LDT, IGHV == "M") 
corRes.M <- cor.test(log2(LDT.M$doubling.time), LDT.M$Factor4)
corRes.M
```

Only U-CLL
```{r}
LDT.U <- filter(LDT, IGHV == "U") 
corRes.U <- cor.test(log2(LDT.U$doubling.time), LDT.U$Factor4)
corRes.U
```

Scatter plot of correlations, stratified by IGHV
```{r}
annoM <- sprintf("'M-CLL: n = %s, coefficient = %1.2f,'~italic(P)~'= %s'",nrow(LDT.M), corRes.M$estimate, formatNum(corRes.M$p.value, digits = 1, format = "e"))
annoU <- sprintf("'U-CLL: n = %s, coefficient = %1.2f,'~italic(P)~'= %s'", nrow(LDT.U), corRes.U$estimate,formatNum(corRes.U$p.value, digits = 1, format = "e"))

corPlot.IGHV <- ggplot(LDT, aes(x = Factor4, y = doubling.time/30, fill = IGHV, col = IGHV)) + 
  geom_point(shape=21, size=3, col = "black") + 
  geom_smooth(method = "lm", se=FALSE, linetype ="dashed" ) + 
  annotate("text", x = Inf, y = Inf, label = annoM,
           hjust=1.05, vjust =1.2, size = 4, parse = TRUE, color = colList[1]) +
  annotate("text", x = Inf, y = Inf, label = annoU,
           hjust=1.05, vjust =2.5, size = 4, parse = TRUE, color = colList[2]) +
  ylab("doubling time (months)") + ggtitle("Lymphocyte doubling time") +
  scale_fill_manual(values = c(M = colList[1],U=colList[2])) +
  scale_color_manual(values = c(M = colList[1],U=colList[2])) +
  scale_y_log10() +
  theme_full + theme(legend.position = "none")

corPlot.IGHV
```

Correlations in untreated patients only
```{r}
LDT.untreat <- doublingTime %>% mutate(Factor4 = facTab[match(patID, facTab$sample),]$value, 
                                       pretreat = demographic[match(patID, demographic$patientID),]$pretreat) %>%
  filter(!is.na(Factor4),!is.na(pretreat)) %>% filter(pretreat == 0)

corRes <- cor.test(LDT.untreat$doubling.time, LDT.untreat$Factor4)

annoText <- sprintf("'coefficient = %1.2f, '~italic(P)~'= %s'",corRes$estimate,formatNum(corRes$p.value, digits = 1, format = "e"))
corPlot.untreat <- ggplot(LDT.untreat, aes(x = Factor4, y = doubling.time/30)) + 
  geom_point(fill =colList[5], shape=21, size=3) + 
  geom_smooth(method = "lm", se=FALSE, color = "grey50", linetype ="dashed" ) + 
  annotate("text", x = 2.2, y = Inf, label = annoText,
           hjust=1, vjust =2, size = 4, parse = TRUE, col= colList[1]) +
  ylab("doubling time (months)") + ggtitle(sprintf("Lymphocyte doubling time\n(untreated patients only, n=%s)",nrow(LDT.untreat))) +
  scale_y_log10() +
  theme_full

corPlot.untreat
```
Variance expalined for lymphocyte doubling time

```{r}
onlyIGHV <- summary(lm(log2(doubling.time) ~ IGHV, data = LDT))
onlyF4 <- summary(lm(log2(doubling.time) ~ Factor4, data = LDT))
combined <- summary(lm(log2(doubling.time) ~ Factor4 + IGHV, data = LDT))
plotTab <- tibble(model = c("IGHV only","Factor4 only", "IGHV + Factor4"),
                  R2 = c(onlyIGHV$adj.r.squared, onlyF4$r.squared, combined$adj.r.squared)) %>%
  mutate(model = factor(model, levels = model))
explainedDT <- ggplot(plotTab, aes(x=model, y = R2)) + geom_bar(stat = "identity", aes(fill = model), width = 0.8) +
  coord_flip(expand = FALSE, xlim = c(0.5,3.5)) + theme_half + scale_fill_manual(values = colList) +
  geom_text(aes(x = model, y =0.01, label = model), hjust =0, fontface = "bold", size =5) +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank(), legend.position = "none",
        axis.title.y = element_text( size =13)) +
  xlab("Predictors") + ylab("Variance explained") 
explainedDT
```
Gene enrichment analysis(Reactome GS)

```{r}
#F1 enrichment positive weights
res.positive <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA",
                               factor = 1,
                               sign = "positive"
)
plot_enrichment_detailed(
  enrichment.results = res.positive,
  factor = 1, 
  max.pathways = 10
)
```

F1 enrichment negative weights
```{r}
res.negative <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA", factor = 1,
                               sign = "negative"
)
plot_enrichment_detailed(
  enrichment.results = res.negative,
  factor = 1, 
  max.pathways = 10)
```


```{r}
```


```{r}
```


```{r}
```


```{r}
```


```{r}
```

\`\`\`

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
