---
title: "RNA-seq data analysis"
output: html_notebook
---

```{r}

```

Different expression analysis (DESeq2)

```{r}
library(DESeq2)
library(dplyr)
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(ggfortify)

load("CLLdata/Kwok_data/rc_Kwok.Rdata")
load("CLLdata/Kwok_data/metadata_Kwok1.Rdata")

rc_Kwok <- rc_Kwok %>% remove_rownames %>% column_to_rownames(var="Name")
rc_Kwok = rc_Kwok[,-1]
```

Check that sample names match in both files

```{r}
all(colnames(rc_Kwok) %in% rownames(metadata_Kwok))
all(colnames(rc_Kwok) == rownames(metadata_Kwok))
```

Create DESeq2Dataset object

```{r}
dds <- DESeqDataSetFromMatrix(countData = rc_Kwok, colData = metadata_Kwok, design = ~ CLLtype)
```

Pre-filter the genes which have low counts

```{r}
dds <- dds[rowSums(counts(dds)) >= 10,]
```

Perform different expression analysis

```{r}
dds <- DESeq2::DESeq(dds)
```

Plotting dispersion estimates

```{r}
DESeq2::plotDispEsts(dds, main="Dispersion Estimates")
```

```{r}
DESeq2::resultsNames(dds)
```

```{r}
deseq2Res <- DESeq2::results(dds, contrast=list("CLLtype_Regression_vs_Indolent", "CLLtype_Progressive_vs_Indolent"), pAdjustMethod = "BH", tidy = TRUE)
deseq2Res <- deseq2Res %>% remove_rownames %>% column_to_rownames(var="row")
deseq2Res <- deseq2Res[order(deseq2Res$padj),]
```

Data visualization

```{r}
# Create the new column
deseq2Res <- deseq2Res %>% mutate(sig=padj<0.01)

# How many of each?
deseq2Res %>% 
  group_by(sig) %>% 
  summarize(n=n())
```

```{r}
#Filtering DE genes
diff_genes = deseq2Res %>% 
  as.data.frame() %>% 
  filter(padj < 0.01) %>% 
  arrange(desc(log2FoldChange), 
          desc(padj))
```

Volcano plot

```{r}
p = ggplot2::ggplot(deseq2Res, ggplot2::aes(log2FoldChange, -log10(pvalue))) +
  ggplot2::geom_point(ggplot2::aes(col = sig), size = 0.5) +
  ggplot2::scale_color_manual(values = c("red", "purple", "grey")) +
  labs(x="log2(fold change)",
       y="-log10 (adj.p-value)",
       title="Volcano Plot of DESeq2 analysis")  +
  theme_bw()+
  theme(plot.title = element_text(hjust = 0.5), 
        legend.position="right", 
        legend.title = element_blank()) 

p + ggrepel::geom_text_repel(data=deseq2Res[1:10, ], ggplot2::aes(label=rownames(deseq2Res[1:10, ])))

```

MA plot

```{r}
deseq2Res1 <- deseq2Res
deseq2Res1$lfcSE <- NULL
deseq2Res1$stat <- NULL
deseq2Res1$pvalue <- NULL
deseq2Res1$sig <- NULL
res2 <- as.data.frame(deseq2Res1) %>% 
  mutate(padj = ifelse(padj <= 0.01, TRUE, FALSE))
```

```{r}
DESeq2::plotMA(res2, main="MA Plot", ylim=c(-3,3))
```

Generating normalized counts

```{r}
dds <- estimateSizeFactors(dds)
sizeFactors(dds)
normalized_counts <- counts(dds, normalized=TRUE)
```

```{r}
dds.vst<-DESeq2::varianceStabilizingTransformation(dds)
exp_Kwok <- SummarizedExperiment::assay(dds.vst)
exp_Kwok <- as.data.frame(exp_Kwok)
```

Selecting DE significat genes

```{r}
diff_genes = deseq2Res %>% 
  as.data.frame() %>% 
  filter(padj < 0.001) %>% 
  arrange(desc(log2FoldChange), 
          desc(padj))
```

```{r}
DE_genes <- merge(exp_Kwok, diff_genes, by = "row.names")
DE_genes = DE_genes %>% 
  as.data.frame() %>% 
  filter(padj < 0.001) %>% 
  arrange(desc(log2FoldChange), 
          desc(padj))

DE_genes <- DE_genes %>% remove_rownames %>% column_to_rownames(var="Row.names")
```

PCA analysis

```{r}
DE_genes1 <- t(DE_genes1)
DE_genes1 <- as.data.frame(DE_genes1)
DE_genes1 <- merge(DE_genes1, metadata_Kwok, by = "row.names")
DE_genes1 <- DE_genes1 %>% remove_rownames %>% column_to_rownames(var="Row.names")
```

```{r}
pca <- prcomp(DE_genes1[,c(1:262)])
summary(pca)
```

```{r}
autoplot(pca, data = DE_genes1, colour = "CLLtype", main = "PCA plot")
```

PCA3D

```{r}
library(rgl)
library(pca3d)
```

```{r}
pca3d(pca, group = DE_genes1[,263] )
```

```{r}
pca3d(pca, group = DE_genes1[,263], 
    fancy= TRUE, bg= "black", 
    show.group.labels=TRUE,
    axes.color= "white", new= TRUE)
```

Heatmap of the most significant genes

```{r}
library("ComplexHeatmap")
library(RColorBrewer)
library(circlize)
library(grid)
```

```{r}
#Heatmap of DE genes (all genes_262, padj<0.001)
ann_col_info <- as.data.frame(metadata_Kwok)
anno_info_colors = list(
  CLLtype = c(Regression = "lightgrey", 
               Indolent = "black", Progressive = "red")
)
```


```{r}
pheatmap(DE_genes, 
         cluster_rows = FALSE,                       
         cluster_cols = TRUE, 
         show_rownames = FALSE, 
         show_colnames = FALSE,
         scale = "row",
         annotation_col = ann_col_info,
         annotation_colors = anno_info_colors,
         main = "Heatmap of different gene expression")
```
Heatmap DE genes Top 30
```{r}
genesTop30 <- DE_genes
genesTop30 <- head(genesTop30, 30)
```

```{r}
genesTop30 <- as.matrix(genesTop30)
```

```{r}
pheatmap(genesTop30, 
         cluster_rows = FALSE,                       
         cluster_cols = TRUE, 
         show_rownames = TRUE, 
         show_colnames = FALSE,
         scale = "row",
         annotation_col = ann_col_info,
         annotation_colors = anno_info_colors,
         main = "Heatmap of different gene expression (Top 30) ")
```


DATASETS INTERSECTION (Kwok & NC)
```{r}
load("DE_genes.RData")
load("CLLproject/Survival_analysis/rna2.RData")
```

Intersection visualization
```{r}
library(VennDetail)
```

```{r}
ven <- venndetail(list(DEG_Kwok = DE_genes1$Name, DEG_NC = rna2$Name))
```

```{r}
plot(ven, type = "vennpie")
```
```{r}
plot(ven, type = "upset")
```
```{r}
```

Merging 2 datasets

```{r}
NC_Kwok_datasets <- merge(commonGenes_expNC, commonGenes_expKwok, by.x = "Name")
```

PCA analysis
```{r}
NC_Kwok_datasets <-NC_Kwok_datasets  %>% remove_rownames %>% column_to_rownames(var="Name")
NC_Kwok_datasets <- t(NC_Kwok_datasets)
NC_Kwok_datasets <- as.data.frame(NC_Kwok_datasets)
```

```{r}
pca <- prcomp(NC_Kwok_datasets, scale. = TRUE)
```

```{r}
autoplot(pca, data = NC_Kwok_datasets, main = "PCA plot of merged datasets")
```

```{r}
pca3d(pca)
```


