---
title: "Multiomic data integration with MOFA2 model"
output: html_notebook
---

# Prepare MOFA model

```{r,  message=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)

library(MOFA2)
library(reticulate)
library(DESeq2)
library(sva)
library(MultiAssayExperiment)
library(tidyverse)
library(BiocParallel)
options(stringsAsFactors=FALSE)
```

```{r, echo=FALSE }
#set the global ggplot theme
theme_set(theme_bw() + theme(axis.text = element_text(size=12), 
                             axis.title = element_text(size=14),
                             plot.title = element_text(size = 15, hjust =0.5, face="bold")))
```
## Prepare each single omic dataset

Load datasets
```{r}
data("drug","gene","rna", "meth", "exprMat_mofa")
```

```{r}
#RNA sequencing
exprMat4 <- as.matrix(exprMat4)
exprMat4 <- assay(exprMat4)
```
```{r}
#DNA methylation array
methData <- assays(meth)[["beta"]]
nTop = 5000
sds <- genefilter::rowSds(methData)
methData <- methData[order(sds, decreasing = T)[1:nTop],]
```
```{r}
#Genetics
gene <- t(gene)
```
```{r}
#Drug screening
viabMat <- mutate(drug, id = paste0(Drug,"_",concIndex)) %>%
  select(patientID, id, normVal) %>%
  spread(key = patientID, value = normVal) %>%
  data.frame() %>% column_to_rownames("id") %>% 
  as.matrix()
```

### Assemble multiAssayExperiment object

```{r}
#Create object
mofaData <- list(mRNA = exprMat4,
                 Mutations = gene, 
                 Methylation = methData,
                 Drugs = viabMat)
```
```{r}
# Create MultiAssayExperiment object 
mofaData <- MultiAssayExperiment::MultiAssayExperiment(
  experiments = mofaData
)
```
```{r}
#Only keep samples that have at least three assays
useSamples <- MultiAssayExperiment::sampleMap(mofaData) %>%
  as_tibble() %>% group_by(primary) %>% summarise(n= length(assay)) %>%
  filter(n >= 3) %>% pull(primary)
mofaData <- mofaData[,useSamples]
```

```{r}
#Dimensions for each dataset
experiments(mofaData)
```
```{r}
#How many samples have the complete datasets
table(table(sampleMap(mofaData)$primary))
```
## Build MOFA object

Build MOFA object from multiAssayExperiment object
```{r, message=FALSE }
MOFAobject <- create_mofa(mofaData)
```
```{r}
MOFAobject
```
Overview of data
```{r, fig.width=8, fig.height=6}
plot_data_overview(MOFAobject)
```
## Setup MOFA training parameters

Define data options
```{r}
data_opts <- get_default_data_options(MOFAobject)
data_opts
```
Define model options
```{r}
model_opts <- get_default_model_options(MOFAobject)
model_opts$num_factors <- 30
model_opts
```
Define training options
```{r}
train_opts <- get_default_training_options(MOFAobject)
train_opts$convergence_mode <- "slow"
train_opts$seed <- 42
train_opts$maxiter <- 5000
train_opts$drop_factor_threshold <- 0.02
train_opts
```
## Run MOFA model 
```{r, eval=FALSE}
MOFAobject <- prepare_mofa(
  object = MOFAobject,
  data_options = data_opts,
  model_options = model_opts,
  training_options = train_opts
)
```
```{r}
MOFAobject <- run_mofa(MOFAobject)
```

#Analysis MOFA model and correlate MOFA factors with clinical outcomes
##Overview of MOFA output
```{r}
library(survival)
library(survminer)
library(maxstat)
library(gridExtra)
library(car)
library(cowplot)
library(egg)
```
Load datasets
```{r}
data("MOFA_CLL", "survival", "gene", "demographic", "doublingTime", "reactomeGS2", "rna")
```
```{r, fig.width=5, fig.height=6}
plot_factor_cor(MOFAobject)
```

### Variance explained by MOFA for each omic data
```{r, fig.width=4.5, fig.height=6}
# Calculate the variance explained (R2) per factor in each view 
r2 <-calculate_variance_explained(MOFAobject)
r2$r2_total
```
```{r}
r2$r2_per_factor
```
```{r, fig.width=6, fig.height=5}
#Total variance explained per view
plot_variance_explained(MOFAobject, plot_total = T)[[2]]
```
```{r, fig.width=8, fig.height=5}
#Variance explained per factor
varExpPlot <- plot_variance_explained(MOFAobject, censor = 0.25)
varExpPlot
```
Get weights and factors

```{r}
allWeights <- get_weights(MOFAobject,
                         views = "all",
                         factors = "all",
                         as.data.frame = TRUE) %>% as_tibble() %>%
  mutate(feature = ifelse(feature == "IGHV.status","IGHV",feature),
         factor = gsub("LF","F",factor))
```
```{r}
allFactors <- get_factors(
  MOFAobject, 
  factors = "all",
  as.data.frame = TRUE
) %>% as_tibble() %>%
  mutate(factor = gsub("LF","F", factor))
```
```{r}
patAnno <- gene[c("IGHV", "trisomy12")] %>% data.frame() %>%
  rownames_to_column("sample") %>%
  filter(!is.na(IGHV),!is.na(trisomy12)) %>%
  mutate(IGHV = ifelse(IGHV==1, "M","U"),
         trisomy12 = ifelse(trisomy12 ==1, "yes","no"))
```
```{r}
allFactors <- left_join(allFactors, 
                        patAnno,
                        by = "sample")
```

### The first two latent factors represents IGHV and trisomy12 

Factor1 profile
```{r, fig.width=7, fig.height=5}
plot_factor(MOFAobject, 
            factors = 1, 
            color_by = "Factor1"
            )
```
```{r, fig.width=7, fig.height=5}
#Factor1 Somatic Mutations
plot_weights(MOFAobject,
             view = "Mutations",
             factor = 1,
             nfeatures = 10,     # Top number of features to highlight
             scale = T           # Scale weights from -1 to 1
)
```
```{r, fig.width=7, fig.height=5}
#Loadings of genomic variations
plot_top_weights(MOFAobject,
                 view = "Mutations",
                 factor = 1,
                 nfeatures = 10,     # Top number of features to highlight
                 scale = T           # Scale weights from -1 to 1
)
```
```{r}
#Heatap for the clusters of Factor1 gene expression
plot_data_heatmap(MOFAobject, 
                  view = "mRNA",
                  factor = 1,  
                  features = 25,
                  denoise = TRUE,
                  cluster_rows = FALSE, cluster_cols = FALSE,
                  show_rownames = TRUE, show_colnames = FALSE,
                  scale = "row"
)
```

###Associations between Factor 1 and epigenetic subtypes
```{r, fig.width=7, fig.height=5}
#Methylation profile
plotTab <- filter(allFactors, factor %in% c("Factor1", "Factor2")) %>%
  mutate(Epitype = methCluster[sample]) %>%
  spread(key = factor ,value = value)

violinLF1_Meth <- ggplot(filter(plotTab, !is.na(Epitype)), aes(x=Epitype, y=Factor1, fill = Epitype)) +
  geom_violin() + geom_point() + scale_fill_manual(values = colList[4:length(colList)]) +
  theme_full + theme(legend.position = "none", axis.title.y = element_text(vjust=-2)) +
  xlab("") + ggtitle("Epigenetic subtypes")
violinLF1_Meth
```

Factor2 profile
```{r, fig.width=7, fig.height=5}
plot_top_weights(MOFAobject,
                 view = "Mutations",
                 factor = 2,
                 nfeatures = 10,     # Top number of features to highlight
                 scale = T )
```
Plot the separation of the samples by the first two factors
```{r, fig.width=7, fig.height=5}
plotTab <- filter(allFactors, factor %in% c("Factor1","Factor2")) %>%
  spread(key = factor, value = value) %>% mutate(trisomy12 = factor(trisomy12)) %>%
  filter(!is.na(IGHV), !is.na(trisomy12))

pcaLF1 <- ggplot(plotTab, aes(x=Factor1, y=Factor2, color = trisomy12, 
                              shape = IGHV, label = sample)) + 
  geom_point(size=3) +
  scale_shape_manual(values = c(M = 16, U =1)) +
  scale_color_manual(values = c(no = colList[1], yes = colList[2])) +
  theme_full + ggtitle("F1 vs F2")
pcaLF1
```
###Factor3 Profile
```{r}
plot_data_heatmap(MOFAobject, 
                  view = "mRNA",
                  features = 30,
                  factor = 3,  
                  denoise = TRUE,
                  cluster_rows = FALSE, cluster_cols = FALSE,
                  show_rownames = TRUE, show_colnames = FALSE,
                  scale = "row")
```
###Factor5 profile
```{r}
plot_data_heatmap(MOFAobject, 
                  view = "mRNA",
                  features = 20,
                  factor = 5, 
                  denoise = TRUE,
                  cluster_rows = FALSE, cluster_cols = TRUE,
                  show_rownames = TRUE, show_colnames = FALSE,
                  scale = "row")
```
Assocations with RNAseq batch

```{r}
allFactors <- get_factors(
  MOFAobject, 
  factors = "all",
  as.data.frame = TRUE
) %>% as_tibble() %>%
  mutate(factor = gsub("LF","F", factor))

batchTab <- filter(allFactors , factor == "Factor5") %>%
  mutate(batch = rna[,match(sample,colnames(rna))]$batch) %>% 
  filter(!is.na(batch)) %>%
  mutate(batch = paste0("batch", batch+1))
```
```{r,fig.width=7, fig.height=5 }
pval <- car::Anova(lm(value ~ factor(batch), batchTab))$`Pr(>F)`[1]
pval <- formatNum(pval, digits = 2)
pAnno <- bquote(italic("P")~"="~.(pval))
colListNew <- colList[-4]
pL4 <- ggplot(batchTab, aes(x=batch, y = value, col = batch)) +
  geom_boxplot() + 
  ggbeeswarm::geom_beeswarm() + 
  scale_color_manual(values = colListNew) +
  annotate("text", x=Inf, y=Inf, label=pAnno, hjust=1.5, vjust=1.5)+
  theme_full +
  theme(legend.position = "none") +
  ylab("F5 value") + xlab("") + ggtitle("F5 ~ RNAseq batch")
pL4
```
# Associations of latent factors to clinical behaviors

## Univariate Cox regression
```{r}
testTab <- left_join(allFactors, survival, by = c(sample = "patientID"))

#for OS
resOS <- filter(testTab, !is.na(OS)) %>%
  group_by(factor) %>%
  do(comSurv(.$value, .$OS, .$died, TRUE)) %>% ungroup() %>%
  arrange(p) %>% mutate(p.adj = p.adjust(p, method = "bonferroni")) %>%
  mutate(Endpoint = "OS")

#for TTT
resTTT <- filter(testTab, !is.na(TTT)) %>%
  group_by(factor) %>%
  do(comSurv(.$value, .$TTT, .$treatedAfter, TRUE)) %>% ungroup() %>%
  arrange(p) %>% mutate(p.adj = p.adjust(p, method = "bonferroni")) %>%
  mutate(Endpoint = "TTT")
```

**Overall survival (OS)**
```{r}
resOS
```
**Time to treatment (TTT)**
```{r}
resTTT
```
**Plot p values and hazard ratios**
```{r}
plotTab <- bind_rows(resOS, resTTT) %>%
  filter(factor %in% c("Factor1","Factor2","Factor3", "Factor4", "Factor5", "Factor6", "Factor7", "Factor8"))

haPlot <- ggplot(plotTab, aes(x=factor, y = HR, col = Endpoint, dodge = Endpoint)) + 
  geom_hline(yintercept = 1, linetype = "dotted") +
  geom_point(position = position_dodge(width=0.8)) +
  geom_errorbar(position = position_dodge(width =0.8), 
                aes(ymin = lower, ymax = higher), width = 0.3, size=1) + 
  geom_text(position = position_dodge2(width = 0.8),
            aes(x=as.numeric(as.factor(factor))+0.15,
                label = sprintf("italic(P)~'='~'%s'",
                                formatNum(p))),
            color = "black",size =4, parse =TRUE) +
  xlab("Factor") + ylab("Hazard ratio") +
  scale_y_log10(limits = c(0.5,4)) +
  coord_flip() + theme_full + theme(legend.title = element_blank(),
                                    legend.position = c(0.2,0.1),
                                    legend.background = element_blank(),
                                    legend.key.size = unit(0.5,"cm"),
                                    legend.key.width = unit(0.6,"cm"),
                                    legend.text = element_text(size=rel(1.2))) +
  ggtitle("Hazard rations of Factors to OS and TTT") +
  scale_color_manual(values = c(OS = colList[3], TTT = colList[5], TFT = colList[4])) 
```
```{r, fig.width=7, fig.height=6}
haPlot
```
## Kaplan-Meiler plots

### KM plot for overall survival (OS)

```{r}
facList <- sort(filter(resOS, p<=0.05)$factor)
osList <- lapply(facList, function(x) {
  eachTab <- filter(testTab, factor == x) %>%
    select(value, OS, died) %>% filter(!is.na(OS))
  pval <- filter(resOS, factor == x)$p
  km(eachTab$value, eachTab$OS, eachTab$died, sprintf("%s VS OS time", x),
     stat = "maxstat", pval = pval, showTable = TRUE)
})
```
```{r}
grid.arrange(grobs = osList, ncol = 2)
```
### KM plot for time to treatment (TTT)
```{r}
facList <- sort(filter(resTTT, p<=0.01)$factor)
tttList <- lapply(facList, function(x) {
  eachTab <- filter(testTab, factor == x) %>%
    select(value, TTT, treatedAfter) %>% filter(!is.na(TTT))
  pval <- filter(resTTT, factor == x)$p
  km(eachTab$value, eachTab$TTT, eachTab$treatedAfter, sprintf("%s VS TTT", x), stat = "maxstat",
     maxTime = 7, pval = pval, showTable = TRUE)
})
```
```{r, fig.width=5, fig.height=6}
tttList
```
### KM plot for subgroup defined by IGHV status and median latent factor values

```{r}
groupTab <- filter(testTab, factor == "Factor4", !is.na(value), !is.na(IGHV)) %>%
  mutate(subgroup = ifelse(value > median(value), paste0(IGHV,"_highFactor4"), paste0(IGHV,"_lowFactor4"))) 

plotList <- list()

# TTT
plotList[["TTT"]] <- km(groupTab$subgroup, groupTab$TTT, groupTab$treatedAfter, "F4 vs TTT", stat = "binary", maxTime = 7, showP = TRUE, showTable = TRUE, yLabelAdjust = -10)
```
```{r}
# OS
plotList[["OS"]] <- km(groupTab$subgroup, groupTab$OS, groupTab$died, "F4 vs OS", stat = "binary", maxTime = 7, showP = TRUE, showTable = TRUE, yLabelAdjust = -10)
```
```{r, fig.height=6.5, fig.width=10}
grid.arrange(grobs = plotList, ncol = 2)
```
```{r}
groupTab <- filter(testTab, factor == "Factor6", !is.na(value), !is.na(IGHV)) %>%
  mutate(subgroup = ifelse(value > median(value), paste0(IGHV,"_highFactor6"), paste0(IGHV,"_lowFactor6"))) 

plotList <- list()

# TTT
plotList[["TTT"]] <- km(groupTab$subgroup, groupTab$TTT, groupTab$treatedAfter, "F6 vs TTT", stat = "binary", maxTime = 7, showP = TRUE, showTable = TRUE, yLabelAdjust = -10)
```
```{r, fig.height=6.5, fig.width=6}
plotList[["TTT"]]
```

## Multi-variate Cox regression for Factor 4 (F4)

Prepare data for multi-variate Cox regression
```{r}
survTab <- survival
facTab <- filter(allFactors, factor == "Factor4")
riskTab <- gene[,c("IGHV","TP53","NOTCH1","del17p","SF3B1")] %>% 
  data.frame() %>% rownames_to_column("patientID") %>%
  mutate(`TP53.del17p` = as.numeric(TP53 | del17p)) %>%
  select(-TP53, -del17p) %>%
  mutate_if(is.numeric, as.factor) %>%
  left_join(select(demographic, patientID, age, sex), by = "patientID") %>%
  mutate(age = age/10) %>%
  mutate(F4 = facTab[match(patientID, facTab$sample),]$value,
         IGHV = factor(IGHV, levels = c(0,1))) %>%
  dplyr::rename(IGHV_mutated = IGHV, Sex_male = sex, Age = age) %>%
  filter(!is.na(F4))
```

**Time to treatment**
```{r}
resTTT <- runCox(survTab, riskTab, "TTT", "treatedAfter")
#summary(surv1)
haTTT <- plotHazard(resTTT, title = "F4 vs TTT") +
  scale_y_log10(limits=c(0.2,5))
```
```{r, fig.height=5, fig.width=6}
haTTT
```
**Overall survival**

```{r}
resOS <- runCox(survTab, riskTab, "OS", "died")
#summary(surv1)
haOS <- plotHazard(resOS, title = "F4 vs OS") +
  scale_y_log10(limits=c(0.2,5))
```
```{r, fig.height=5, fig.width=6 }
haOS
```
```{r}
survTab <- survival
facTab <- filter(allFactors, factor == "Factor6")
riskTab <- gene[,c("IGHV","TP53","NOTCH1","del17p","SF3B1")] %>% 
  data.frame() %>% rownames_to_column("patientID") %>%
  mutate(`TP53.del17p` = as.numeric(TP53 | del17p)) %>%
  select(-TP53, -del17p) %>%
  mutate_if(is.numeric, as.factor) %>%
  left_join(select(demographic, patientID, age, sex), by = "patientID") %>%
  mutate(age = age/10) %>%
  mutate(F6 = facTab[match(patientID, facTab$sample),]$value,
         IGHV = factor(IGHV, levels = c(0,1))) %>%
  dplyr::rename(IGHV_mutated = IGHV, Sex_male = sex, Age = age) %>%
  filter(!is.na(F6))
```
```{r, fig.height=5, fig.width=6}
resTTT <- runCox(survTab, riskTab, "TTT", "treatedAfter")
#summary(surv1)
haTTT <- plotHazard(resTTT, title = "F6 vs TTT") +
  scale_y_log10(limits=c(0.2,5))

haTTT
```
## Correlation between F4 and demographics
**Pretreatment**
```{r, fig.height=5, fig.width=5}
plotTab <- left_join(demographic, facTab, by = c(patientID = "sample")) %>%
  filter(!is.na(value)) %>%
  mutate(pretreat = ifelse(is.na(pretreat),NA,
                           ifelse(pretreat ==1 , "yes","no")),
         sex = ifelse(is.na(sex),NA,
                           ifelse(sex =="f" , "Female","Male")))
corRes <- t.test(value ~ pretreat, plotTab)
annoP <- paste("italic(P)~'='~",formatNum(corRes$p.value, digits = 1, format = "e"))

plotTab <- filter(plotTab, !is.na(pretreat)) %>%
  group_by(pretreat) %>% mutate(n=n()) %>% ungroup() %>%
  mutate(pretreat = sprintf("%s\n(n=%s)",pretreat,n))

plotTreat <- ggplot(filter(plotTab,!is.na(pretreat)), aes(x = pretreat, y = value)) + 
  geom_violin(aes(fill = pretreat)) +
  geom_point() + 
  annotate("text", x = -Inf, y = Inf, label = annoP,
           hjust=-0.2, vjust =2, size = 5, parse = TRUE, col= colList[1]) +
  scale_fill_manual(values = colList[3:length(colList)]) +
  ylab("F4") + xlab("Pretreatment") +
  theme_full + theme(legend.position = "none")
plotTreat
```
## Association between F4 and outcomes in treatment-naive patients

### Univariate Cox regression
```{r}
survival.untreated <- survival %>% mutate(pretreat = demographic[match(patientID, demographic$patientID),]$pretreat) %>%
  filter(pretreat ==0)
testTab <- left_join(allFactors, survival.untreated, by = c(sample = "patientID"))

#for OS
resOS <- filter(testTab, !is.na(OS), factor == "Factor4") %>%
  group_by(factor) %>%
  do(comSurv(.$value, .$OS, .$died, TRUE)) %>% ungroup() %>%
  arrange(p) %>% mutate(p.adj = p.adjust(p, method = "BH")) %>%
  mutate(Endpoint = "OS")

resOS
```
```{r}
#for TTT
resTTT <- filter(testTab, !is.na(TTT), factor == "Factor4") %>%
  group_by(factor) %>%
  do(comSurv(.$value, .$TTT, .$treatedAfter, TRUE)) %>% ungroup() %>%
  arrange(p) %>% mutate(p.adj = p.adjust(p, method = "BH")) %>%
  mutate(Endpoint = "TTT")
resTTT
```
### Kaplan-Meiler plots

KM plot for overall survival (OS)

```{r}
#KM plot for overall survival (OS)
eachTab <- filter(testTab, factor == "Factor4") %>%
  select(value, OS, died) %>% filter(!is.na(OS))
pval <- filter(resOS, factor == "Factor4")$p
kmOS.untreat <- km(eachTab$value, eachTab$OS, eachTab$died, "F4 VS OS time",
                   stat = "maxstat", pval = pval, showTable = TRUE)
```
```{r,fig.width=5, fig.height=6 }
kmOS.untreat
```
KM plot for time to treatment (TTT)
```{r}
eachTab <- filter(testTab, factor == "Factor4") %>%
  select(value, TTT, treatedAfter) %>% filter(!is.na(TTT))
pval <- filter(resTTT, factor == "Factor4")$p
kmTTT.untreat <- km(eachTab$value, eachTab$TTT, eachTab$treatedAfter, "F4 VS TTT",
                    stat = "maxstat", pval = pval, showTable = TRUE)
```
```{r,fig.width=5, fig.height=6 }
kmTTT.untreat
```
## Multi-variate Cox regression

Prepare data for multi-variate Cox regression

```{r}
survTab <- survival.untreated
facTab <- filter(allFactors, factor == "Factor4")
riskTab <- gene[,c("IGHV","TP53","del17p","SF3B1")] %>% 
  data.frame() %>% rownames_to_column("patientID") %>%
  mutate(`TP53.del17p` = as.numeric(TP53 | del17p)) %>%
  select(-TP53, -del17p) %>%
  mutate_if(is.numeric, as.factor) %>%
  left_join(select(demographic, patientID, age, sex), by = "patientID") %>%
  mutate(age = age/10) %>%
  mutate(Factor4 = facTab[match(patientID, facTab$sample),]$value,
         IGHV = factor(IGHV, levels = c(0,1))) %>%
  dplyr::rename(IGHV_mutated = IGHV, Sex_male = sex, Age = age) %>%
  filter(!is.na(Factor4))
```

**Time to treatment**
```{r, fig.width=6, fig.height=5}
resTTT <- runCox(survTab, riskTab, "TTT", "treatedAfter")

#summary(surv1)
haTTT.untreat <- plotHazard(resTTT, title = "F4 vs TTT ")
haTTT.untreat
```
**Overall survival**
```{r, fig.width=6, fig.height=5}
resOS <- runCox(survTab, riskTab, "OS", "died")

#summary(surv1)
haOS.untreat <- plotHazard(resOS,"F4 vs OS")
haOS.untreat
```
## Correlation between F4 and Lymphocyte doubling time

### Univariate test

Pearson's correlation test
```{r}
LDT <- doublingTime %>% mutate(Factor4 = facTab[match(patID, facTab$sample),]$value) %>%
  filter(!is.na(Factor4)) %>%
  mutate(IGHV = as.factor(facTab[match(patID, facTab$sample),]$IGHV))
corRes <- cor.test(log10(LDT$doubling.time), LDT$Factor4)
```
Scatter plot of correlations
```{r, fig.width=6, fig.height=5}
pval <- formatNum(corRes$p.value, digits = 1, format = "e")
annoN <- sprintf("n = %s", nrow(LDT))
annoP <- bquote(italic("P")~"="~.(pval))
annoCoef <- sprintf("coefficient = %1.2f",corRes$estimate)

corPlot <- ggplot(LDT, aes(x = Factor4, y = doubling.time/30)) + 
  geom_point(fill =colList[5], shape =21, size=3) + 
  geom_smooth(method = "lm", se=FALSE, color = "grey50", linetype ="dashed" ) +
  annotate("text", x = max(LDT$Factor4), y = Inf, label = annoN,
           hjust=1, vjust =1.5, size = 5, parse = FALSE, col= colList[1]) +
  annotate("text", x = max(LDT$Factor4), y = Inf, label = annoP,
           hjust=1, vjust =3.5, size = 5, parse = FALSE, col= colList[1]) +
  annotate("text", x = max(LDT$Factor4), y = Inf, label = annoCoef,
           hjust=1, vjust =5.5, size = 5, parse = FALSE, col= colList[1]) +
  ylab(bquote("doubling time (months)")) + ggtitle("Lymphocyte doubling time") +
  scale_y_log10() +
  theme_full

corPlot
```
```{r,fig.width=6, fig.height=5 }
facTab <- filter(allFactors, factor == "Factor6")
LDT <- doublingTime %>% mutate(Factor6 = facTab[match(patID, facTab$sample),]$value) %>%
  filter(!is.na(Factor6)) %>%
  mutate(IGHV = as.factor(facTab[match(patID, facTab$sample),]$IGHV))
corRes <- cor.test(log10(LDT$doubling.time), LDT$Factor6)
```
```{r, fig.width=6, fig.height=6}
pval <- formatNum(corRes$p.value, digits = 1, format = "e")
annoN <- sprintf("n = %s", nrow(LDT))
annoP <- bquote(italic("P")~"="~.(pval))
annoCoef <- sprintf("coefficient = %1.2f",corRes$estimate)

corPlot <- ggplot(LDT, aes(x = Factor6, y = doubling.time/30)) + 
  geom_point(fill =colList[5], shape =21, size=3) + 
  geom_smooth(method = "lm", se=FALSE, color = "grey50", linetype ="dashed" ) +
  annotate("text", x = max(LDT$Factor6), y = Inf, label = annoN,
           hjust=1, vjust =1.5, size = 5, parse = FALSE, col= colList[1]) +
  annotate("text", x = max(LDT$Factor6), y = Inf, label = annoP,
           hjust=1, vjust =3.5, size = 5, parse = FALSE, col= colList[1]) +
  annotate("text", x = max(LDT$Factor6), y = Inf, label = annoCoef,
           hjust=1, vjust =5.5, size = 5, parse = FALSE, col= colList[1]) +
  ylab(bquote("doubling time (months)")) + ggtitle("Lymphocyte doubling time") +
  scale_y_log10() +
  theme_full

corPlot
```
### Consider IGHV status
#### IGHV as a covariate

ANOVA test

```{r}
LDT <- doublingTime %>% mutate(Factor6 = facTab[match(patID, facTab$sample),]$value,
                               IGHV = as.factor(facTab[match(patID, facTab$sample),]$IGHV)) %>%
  filter(!is.na(Factor6),!is.na(IGHV))

corRes <- car::Anova(lm(log10(doubling.time) ~ Factor6 + IGHV, data = LDT))
corRes
```

#### Only M-CLL
```{r}
LDT.M <- filter(LDT, IGHV == "M") 
corRes.M <- cor.test(log2(LDT.M$doubling.time), LDT.M$Factor6)
corRes.M
```

#### Only U-CLL
```{r}
LDT.U <- filter(LDT, IGHV == "U") 
corRes.U <- cor.test(log2(LDT.U$doubling.time), LDT.U$Factor6)
corRes.U
```

Scatter plot of correlations, stratified by IGHV
```{r, fig.width=7, fig.height=6}
annoM <- sprintf("'M-CLL: n = %s, coefficient = %1.2f,'~italic(P)~'= %s'",nrow(LDT.M), corRes.M$estimate, formatNum(corRes.M$p.value, digits = 1, format = "e"))
annoU <- sprintf("'U-CLL: n = %s, coefficient = %1.2f,'~italic(P)~'= %s'", nrow(LDT.U), corRes.U$estimate,formatNum(corRes.U$p.value, digits = 1, format = "e"))

corPlot.IGHV <- ggplot(LDT, aes(x = Factor6, y = doubling.time/30, fill = IGHV, col = IGHV)) + 
  geom_point(shape=21, size=3, col = "black") + 
  geom_smooth(method = "lm", se=FALSE, linetype ="dashed" ) + 
  annotate("text", x = Inf, y = Inf, label = annoM,
           hjust=1.05, vjust =1.2, size = 4, parse = TRUE, color = colList[1]) +
  annotate("text", x = Inf, y = Inf, label = annoU,
           hjust=1.05, vjust =2.5, size = 4, parse = TRUE, color = colList[2]) +
  ylab("doubling time (months)") + ggtitle("Lymphocyte doubling time") +
  scale_fill_manual(values = c(M = colList[1],U=colList[2])) +
  scale_color_manual(values = c(M = colList[1],U=colList[2])) +
  scale_y_log10() +
  theme_full + theme(legend.position = "none")

corPlot.IGHV
```
Correlations in untreated patients only
```{r, fig.width=7, fig.height=6}
LDT.untreat <- doublingTime %>% mutate(Factor6 = facTab[match(patID, facTab$sample),]$value, 
                                       pretreat = demographic[match(patID, demographic$patientID),]$pretreat) %>%
  filter(!is.na(Factor6),!is.na(pretreat)) %>% filter(pretreat == 0)

corRes <- cor.test(LDT.untreat$doubling.time, LDT.untreat$Factor6)

annoText <- sprintf("'coefficient = %1.2f, '~italic(P)~'= %s'",corRes$estimate,formatNum(corRes$p.value, digits = 1, format = "e"))
corPlot.untreat <- ggplot(LDT.untreat, aes(x = Factor6, y = doubling.time/30)) + 
  geom_point(fill =colList[5], shape=21, size=3) + 
  geom_smooth(method = "lm", se=FALSE, color = "grey50", linetype ="dashed" ) + 
  annotate("text", x = 2.2, y = Inf, label = annoText,
           hjust=1, vjust =2, size = 4, parse = TRUE, col= colList[1]) +
  ylab("doubling time (months)") + ggtitle(sprintf("Lymphocyte doubling time\n(untreated patients only, n=%s)",nrow(LDT.untreat))) +
  scale_y_log10() +
  theme_full

corPlot.untreat
```

### Variance expalined for lymphocyte doubling time
```{r, fig.width=5, fig.height=5}
onlyIGHV <- summary(lm(log2(doubling.time) ~ IGHV, data = LDT))
onlyF4 <- summary(lm(log2(doubling.time) ~ Factor6, data = LDT))
combined <- summary(lm(log2(doubling.time) ~ Factor6 + IGHV, data = LDT))
plotTab <- tibble(model = c("IGHV only","Factor6 only", "IGHV + Factor6"),
                  R2 = c(onlyIGHV$adj.r.squared, onlyF4$r.squared, combined$adj.r.squared)) %>%
  mutate(model = factor(model, levels = model))
explainedDT <- ggplot(plotTab, aes(x=model, y = R2)) + geom_bar(stat = "identity", aes(fill = model), width = 0.8) +
  coord_flip(expand = FALSE, xlim = c(0.5,3.5)) + theme_half + scale_fill_manual(values = colList) +
  geom_text(aes(x = model, y =0.01, label = model), hjust =0, fontface = "bold", size =5) +
  theme(axis.ticks.y = element_blank(), axis.text.y = element_blank(), legend.position = "none",
        axis.title.y = element_text( size =13)) +
  xlab("Predictors") + ylab("Variance explained") 
explainedDT
```

##Association between F6 and genomics
```{r}
#Load datasets (MOFA2_CLL, "gene", "rna", "meth", mutLoad", "tabFACS")
```
Process MOFA model
```{r}
#get factor values
facTab <- get_factors(
  MOFAobject, 
  factors = "Factor6",
  as.data.frame = TRUE
) %>% as_tibble() 

#get all factors
facTab.all <- get_factors(
  MOFAobject,
  factors = "all",
  as.data.frame = TRUE
) %>% as_tibble()

#get weights
weightTab <- get_weights(
  MOFAobject,
  factors = "all",
  as.data.frame = TRUE
) %>% as_tibble() %>%
  mutate(feature = ifelse(feature == "IGHV.status", "IGHV", feature))

```
```{r}
```

###Loading of features in the genomic view
```{r, fig.width=6, fig.height=5}
plotGeneWeight <- plot_top_weights(MOFAobject, view =  "Mutations", factor = "Factor6", nfeatures = 10) + ylab("Absolute loading on F6") +
  theme(plot.margin = margin(10,0,10,0))
plotGeneWeight
```
```{r, fig.width=6, fig.height=5}
plotGeneWeight <- plot_top_weights(MOFAobject, view =  "Mutations", factor = "Factor4", nfeatures = 10) + ylab("Absolute loading on F4") +
  theme(plot.margin = margin(10,0,10,0))
plotGeneWeight
```
###Test of associations between genetic variations with latent factors
Studentâ€™s t-test F6
```{r}
sampleOverlap <- intersect(rownames(gene.unfiltered), facTab$sample)
genData <- gene.unfiltered[sampleOverlap,]
```
```{r}
#remove gene with higher than 40% missing values
genData <- genData[,colSums(is.na(genData))/nrow(genData) <= 0.4]

#remove genes with less than 3 mutated cases
genData <- genData[,colSums(genData, na.rm = TRUE) >= 3]

#t-test
y <- facTab[match(sampleOverlap, facTab$sample),]$value
tRes <- apply(genData, 2, function(x) {
  res <- t.test(y ~ as.factor(x), var.equal = TRUE)
  data.frame(p = res$p.value, 
             df = res$estimate[[2]] - res$estimate[[1]])
}) %>% bind_rows() %>% mutate(gene = colnames(genData),
                              p.adj = p.adjust(p, method = "BH")) %>%
  arrange(p)
filter(tRes, p.adj <0.05) %>% mutate_if(is.numeric, formatNum, digits =3, format = "e") %>% DT::datatable()
```
```{r,fig.width=8, fig.height=7 }
#Volcano plot
plotGeneVolcano <- plotVolcano(tRes, posCol = colList[1], negCol = colList[2],
            x_lab = "Difference of mean", ifLabel = TRUE) + 
  ggtitle("Association between genetic variation and F6") +
  theme(legend.position = "none",
        plot.margin = margin(8,8,8,8),
        axis.title.y = element_text(vjust=0))
plotGeneVolcano
```

Student's t-test F4
```{r}
#get factor values
facTab <- get_factors(
  MOFAobject, 
  factors = "Factor4",
  as.data.frame = TRUE
) %>% as_tibble() 

```
```{r}
sampleOverlap <- intersect(rownames(gene.unfiltered), facTab$sample)
genData <- gene.unfiltered[sampleOverlap,]

#remove gene with higher than 40% missing values
genData <- genData[,colSums(is.na(genData))/nrow(genData) <= 0.4]

#remove genes with less than 3 mutated cases
genData <- genData[,colSums(genData, na.rm = TRUE) >= 3]

#t-test
y <- facTab[match(sampleOverlap, facTab$sample),]$value
tRes <- apply(genData, 2, function(x) {
  res <- t.test(y ~ as.factor(x), var.equal = TRUE)
  data.frame(p = res$p.value, 
             df = res$estimate[[2]] - res$estimate[[1]])
}) %>% bind_rows() %>% mutate(gene = colnames(genData),
                              p.adj = p.adjust(p, method = "BH")) %>%
  arrange(p)
filter(tRes, p.adj <0.05) %>% mutate_if(is.numeric, formatNum, digits =3, format = "e") %>% DT::datatable()
```
Volcano plot
```{r, fig.width=7, fig.height=7}
plotGeneVolcano <- plotVolcano(tRes, posCol = colList[1], negCol = colList[2],
            x_lab = "Difference of mean", ifLabel = TRUE) + 
  ggtitle("Association between genetic variation and F4") +
  theme(legend.position = "none",
        plot.margin = margin(8,8,8,8),
        axis.title.y = element_text(vjust=0))
plotGeneVolcano
```
Heatmap plot showing associated genetic features (5% FDR), arranged by F6 values.

```{r, fig.width=10, fig.height=3}
library(pheatmap)
library(grid) 
annoCol <- data.frame(row.names = sampleOverlap,
                      F6 = facTab[match(sampleOverlap, facTab$sample),]$value)
annoCol <- annoCol[order(annoCol$F6),,drop = FALSE]
plotMat <- t(genData[rownames(annoCol),filter(tRes, p.adj <= 0.05)$gene])
plotMat[is.na(plotMat)] <- 0
breaks <- c(0,0.5,1)
colors <- c("white","black")
rowLabs <- italicizeGene(rownames(plotMat))
plotGeneHeatmap <- pheatmap(plotMat, cluster_cols = FALSE, cluster_rows = FALSE, 
                            annotation_col = annoCol, border_color = "black",
         breaks = breaks, color = colors, show_colnames = FALSE, 
         labels_row = parse(text = rowLabs),
         legend = FALSE, silent = TRUE)$gtable
plot_grid(plotGeneHeatmap)
```

Heatmap plot showing associated genetic features (5% FDR), arranged by F4 values.
```{r, fig.width=10, fig.height=3}
annoCol <- data.frame(row.names = sampleOverlap,
                      F4 = facTab[match(sampleOverlap, facTab$sample),]$value)
annoCol <- annoCol[order(annoCol$F4),,drop = FALSE]
plotMat <- t(genData[rownames(annoCol),filter(tRes, p.adj <= 0.05)$gene])
plotMat[is.na(plotMat)] <- 0
breaks <- c(0,0.5,1)
colors <- c("white","black")
rowLabs <- italicizeGene(rownames(plotMat))
plotGeneHeatmap <- pheatmap(plotMat, cluster_cols = FALSE, cluster_rows = FALSE, 
                            annotation_col = annoCol, border_color = "black",
         breaks = breaks, color = colors, show_colnames = FALSE, 
         labels_row = parse(text = rowLabs),
         legend = FALSE, silent = TRUE)$gtable
plot_grid(plotGeneHeatmap)
```

##Assocation between F4/F6 and DNA methylation
###Correlation test using linear model
```{r}
library(limma)
methMat <- assays(meth[,colnames(meth) %in% facTab$sample])[["beta"]]
factorMatrix <- facTab.all %>% spread(key = factor, value = value) %>%
  data.frame() %>% column_to_rownames("sample") 
factorMatrix <- factorMatrix[colnames(methMat),]
factorMatrix <- factorMatrix[,complete.cases(t(factorMatrix))]
factorMatrix <- factorMatrix[,-1] #delete column
designMat <- model.matrix(~1 + .,factorMatrix)
fit <- lmFit(methMat, designMat)
fit2 <- eBayes(fit)
corRes <- limma::topTable(fit2, number ="all", coef = "Factor6", adjust.method = "BH") %>%
  rownames_to_column("id") %>% as_tibble()
corRes.LF4 <- limma::topTable(fit2, number ="all", coef = "Factor4", adjust.method = "BH") %>%
  rownames_to_column("id") %>% as_tibble()
```
```{r}
#Number of tested associations
dim(methMat)
```
###Compare number of significant correlations in F4 and F6
```{r, fig.width=6, fig.height=5}
numTab.LF6 <- filter(corRes, adj.P.Val <= 0.01) %>% mutate(direction = ifelse(t>0, "pos","neg")) %>%
  group_by(direction) %>% summarise(number = length(id)) %>%
  mutate(factor = "F6")
numTab.LF4 <- filter(corRes.LF4, adj.P.Val <= 0.01) %>% mutate(direction = ifelse(t>0, "pos","neg")) %>%
  group_by(direction) %>% summarise(number = length(id)) %>% 
  mutate(factor = "F4")

plotTab <- bind_rows(numTab.LF4, numTab.LF6)
sigNumPlot <- ggplot(plotTab, aes(x=factor, y=number, fill = direction)) + 
  geom_bar(stat = "identity", width = 0.7,
           position = position_dodge2(width = 6),
           col = "black") +
  geom_text(aes(label=number), 
            position = position_dodge(width = 0.9),
            size=4, hjust=-0.1)  +
  scale_fill_manual(name = "", labels = c("Hypomethylated","Hypermethylated"), values = colList) +
  coord_flip(ylim = c(0,50000), expand = FALSE) + xlab("") + ylab("Number of significant associations") + theme_half +
  theme(legend.position = c(0.75,0.42), legend.text = element_text(size=12), legend.background = element_rect(fill = NA))
sigNumPlot
```

Correlation between mean methylation values and F6
```{r,fig.width=6, fig.height=6 }
meanBeta <- colMeans(methMat)
plotTab <- tibble(patID = names(meanBeta),
                  meanBeta = meanBeta,
                  F6 = facTab[match(patID, facTab$sample),]$value)

corBeta <- cor.test(plotTab$meanBeta, plotTab$F6)

annoCoef <- paste("'coefficient ='~",format(corBeta$estimate,digits = 2))
annoP <- paste("italic(P)~'='~",formatNum(corBeta$p.value, digits = 1, format = "e"))

plotMeanBeta <- ggplot(plotTab, aes(x = F6, y = meanBeta)) + 
  geom_point(shape = 21, fill =colList[3], size=3) + 
  geom_smooth(method = "lm", se=FALSE, color = "grey50", linetype ="dashed" ) + 
  annotate("text", x = 3, y = Inf, label = annoCoef,
           hjust=1, vjust =1.5, size = 5, parse = TRUE, col= colList[1]) +
  annotate("text", x = 3, y = Inf, label = annoP,
           hjust=1, vjust =3, size = 5, parse = TRUE, col= colList[1]) +
  ylab("Mean beta values") + xlab("F6") +
  ggtitle("F6 vs overall DNA methylation level") +
  theme_full 

plotMeanBeta
```
Correlation between mean methylation values and F4
```{r, fig.width=6, fig.height=6}
#get factor values
facTab <- get_factors(
  MOFAobject, 
  factors = "Factor4",
  as.data.frame = TRUE
) %>% as_tibble() 

meanBeta <- colMeans(methMat)
plotTab <- tibble(patID = names(meanBeta),
                  meanBeta = meanBeta,
                  F4 = facTab[match(patID, facTab$sample),]$value)

corBeta <- cor.test(plotTab$meanBeta, plotTab$F4)

annoCoef <- paste("'coefficient ='~",format(corBeta$estimate,digits = 2))
annoP <- paste("italic(P)~'='~",formatNum(corBeta$p.value, digits = 1, format = "e"))

plotMeanBeta <- ggplot(plotTab, aes(x = F4, y = meanBeta)) + 
  geom_point(shape = 21, fill =colList[3], size=3) + 
  geom_smooth(method = "lm", se=FALSE, color = "grey50", linetype ="dashed" ) + 
  annotate("text", x = 3, y = Inf, label = annoCoef,
           hjust=1, vjust =1.5, size = 5, parse = TRUE, col= colList[1]) +
  annotate("text", x = 3, y = Inf, label = annoP,
           hjust=1, vjust =3, size = 5, parse = TRUE, col= colList[1]) +
  ylab("Mean beta values") + xlab("F4") +
  ggtitle("F4 vs overall DNA methylation level") +
  theme_full 

plotMeanBeta
```
##Reactome pathway Enrichment Gsea
```{r}
#F1 enrichment on positive weights
res.positive <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA",
                               factor = 1,
                               sign = "positive")
```
```{r, fig.width=8, fig.height=4}
plot_enrichment(res.positive, factor = "Factor1", max.pathways = 20)+ ggtitle("Reactome enrichment F1 (positive)")
```
F1 enrichment on negative weights
```{r}
res.negative <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA", factor = 1,
                               sign = "negative")
```
```{r, fig.width=9, fig.height=4}
plot_enrichment(res.negative, factor = "Factor1", max.pathways = 10)+ ggtitle("Reactome enrichment F1 (negative)")
```
F2 enrichment on positive weights
```{r}
res.positive <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA",
                               factor = 2,
                               sign = "positive")
```
```{r, fig.width=10, fig.height=4}
plot_enrichment(res.positive, factor = "Factor2", max.pathways = 20)+ ggtitle("Reactome enrichment F2 (positive)")
```
F2 enrichment on negative weights
```{r}
res.negative <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA", factor = 2,
                               sign = "negative")
```
```{r,fig.width=10, fig.height=4 }
plot_enrichment(res.negative, factor = "Factor2", max.pathways = 10)+ ggtitle("Reactome enrichment F2 (negative)")
```

F3 enrichment on positive weights
```{r}
res.positive <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA",
                               factor = 3,
                               sign = "positive")
```
```{r, fig.width=10, fig.height=6 }
plot_enrichment(res.positive, factor = "Factor3", max.pathways = 20)+ ggtitle("Reactome enrichment F3 (positive)")
```
F3 enrichment on negative weights
```{r}
res.negative <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA", factor = 3,
                               sign = "negative")
```
```{r, fig.width=12, fig.height=6 }
plot_enrichment(res.negative, factor = "Factor3", max.pathways = 10)+ ggtitle("Reactome enrichment F3 (negative)")
```
F4 enrichment on positive weights
```{r}
res.positive <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA",
                               factor = 4,
                               sign = "positive")
```
```{r, fig.width=10, fig.height=6 }
plot_enrichment(res.positive, factor = "Factor4", max.pathways = 20)+ ggtitle("Reactome enrichment F4 (positive)")
```
F4 enrichment on negative weights
```{r}
res.negative <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA", factor = 4,
                               sign = "negative")
```
```{r, fig.width=9, fig.height=4}
plot_enrichment(res.negative, factor = "Factor4", max.pathways = 20)+ ggtitle("Reactome enrichment F4 (negative)")
```
F6 enrichment on positive weights
```{r}
res.positive <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA",
                               factor = 6,
                               sign = "positive")
```
```{r, fig.width=9, fig.height=5}
plot_enrichment(res.positive, factor = "Factor6", max.pathways = 20)+ ggtitle("Reactome enrichment F6 (positive)")
```

F7 enrichment on positive weights
```{r}
res.positive <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA",
                               factor = 7,
                               sign = "positive")
```
```{r, fig.width=8, fig.height=4}
plot_enrichment(res.positive, factor = "Factor7", max.pathways = 20)+ ggtitle("Reactome enrichment F7 (positive)")
```
F7 enrichment on negative weights
```{r}
res.negative <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA", factor = 7,
                               sign = "negative")
```
```{r, fig.width=9, fig.height=4 }
plot_enrichment(res.negative, factor = "Factor7", max.pathways = 20)+ ggtitle("Reactome enrichment F7 (negative)")
```
F8 enrichment on positive weights
```{r}
res.positive <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA",
                               factor = 8,
                               sign = "positive")
```
```{r, fig.width=13, fig.height=6}
plot_enrichment(res.positive, factor = "Factor8", max.pathways = 10)+ ggtitle("Reactome enrichment F8 (positive)")
```
F8 enrichment on negative weights
```{r}
res.negative <- run_enrichment(MOFAobject, 
                               feature.sets = reactomeGS3, 
                               view = "mRNA", factor = 8,
                               sign = "negative")
```
```{r, fig.width=13, fig.height=6}
plot_enrichment(res.negative, factor = "Factor8", max.pathways = 10)+ ggtitle("Reactome enrichment F8 (negative)")
```

