---
title: "Kwok RNA-seq data analysis"
output: html_notebook
---

```{r}

```

Different expression analysis (DESeq2)

```{r}
library(DESeq2)
library(dplyr)
library(tidyverse)
library(ggplot2)
library(ggrepel)
library(ggfortify)


#load("CLLdata/Kwok_data/rc_Kwok.Rdata")
#load("CLLdata/Kwok_data/metadata_Kwok1.Rdata")

```

Check that sample names match in both files

```{r}
all(colnames(rc_Kwok) %in% rownames(metadata_Kwok))
all(colnames(rc_Kwok) == rownames(metadata_Kwok))
```

Create DESeq2Dataset object

```{r}
dds <- DESeqDataSetFromMatrix(countData = rc_Kwok, colData = metadata_Kwok, design = ~ CLLtype)
```

Pre-filter the genes which have low counts

```{r}
dds <- dds[rowSums(counts(dds)) >= 10,]
```

Perform different expression analysis

```{r}
dds <- DESeq2::DESeq(dds)
```

Dispersion plot
```{r}
DESeq2::plotDispEsts(dds, main="Dispersion Estimates")
```

Using contrast

```{r}
res1 <- results(dds, contrast=c("CLLtype","Indolent","Progressive"), pAdjustMethod = "BH", tidy = TRUE)
res2 <- results(dds, contrast=c("CLLtype","Regression","Indolent"), pAdjustMethod = "BH", tidy = TRUE)
res3 <- results(dds, contrast=c("CLLtype","Regression","Progressive"), pAdjustMethod = "BH", tidy = TRUE)
```


Ordering results based on padj value
```{r}
res1 <- res1 %>% remove_rownames %>% column_to_rownames(var="row")
res1 <- res1[order(res1$padj),]

res2 <- res2 %>% remove_rownames %>% column_to_rownames(var="row")
res2 <- res2[order(res2$padj),]

res3 <- res3 %>% remove_rownames %>% column_to_rownames(var="row")
res3 <- res3[order(res3$padj),]
```

Getting DE genes
```{r}
#up-regulated (Indolent vs Progressive)
sum(res1$padj < 0.01 & res1$log2FoldChange > 1, na.rm=TRUE)
#down-regulated (Indolent vs Progressive)
sum(res1$padj < 0.01 & res1$log2FoldChange < -1, na.rm=TRUE)
```
```{r}
#up-regulated (Regression vs Indolent)
sum(res2$padj < 0.01 & res2$log2FoldChange > 1, na.rm=TRUE)
#down-regulated (Regression vs Indolent)
sum(res2$padj < 0.01 & res2$log2FoldChange < -1, na.rm=TRUE)

```
```{r}
#up-regulated (Regression vs Progressive)
sum(res3$padj < 0.01 & res3$log2FoldChange > 1, na.rm=TRUE)
#down-regulated (Regression vs Progressive)
sum(res3$padj < 0.01 & res3$log2FoldChange < -1, na.rm=TRUE)
```
```{r}
```
Filtering DE genes (up- & down- regulated)
```{r}
res1Sig_up <- subset(res1, res1$padj < 0.01 & res1$log2FoldChange > 1)
res1Sig_down <- subset(res1, res1$padj < 0.01 & res1$log2FoldChange < -1)

res2Sig_up <- subset(res2, res2$padj < 0.01 & res2$log2FoldChange > 1)
res2Sig_down <- subset(res2, res2$padj < 0.01 & res2$log2FoldChange < -1)

res3Sig_up <- subset(res3, res3$padj < 0.01 & res3$log2FoldChange > 1)
res3Sig_down <- subset(res3, res3$padj < 0.01 & res3$log2FoldChange < -1)
```

Indolent vs Progressive
```{r}
head(res1Sig_up[ order( -res1Sig_up$log2FoldChange ), ], 10)
```
```{r}
head(res1Sig_down[ order( res1Sig_down$log2FoldChange ), ], 10)
```

Regression vs Indolent

```{r}
head(res2Sig_up[ order( -res2Sig_up$log2FoldChange ), ], 10)
```
```{r}
head(res2Sig_down[ order( res2Sig_down$log2FoldChange ), ], 10)
```

EXPLORATORY ANALYSIS AND VISUALIZATION
```{r}
#Volcano plot
library(EnhancedVolcano)

#Indolent vs Progressive

EnhancedVolcano(toptable = res1,            
                x = "log2FoldChange",           
                y = "padj",                     
                lab = rownames(res1),
                pCutoff = 0.01,
                FCcutoff = 1,
                title = "Indolent vs Progressive"
                )
```

```{r}
#Regression vs Indolent
EnhancedVolcano(toptable = res2,            
                x = "log2FoldChange",           
                y = "padj",                     
                lab = rownames(res2),
                pCutoff = 0.01,
                FCcutoff = 1,
                title = "Regression vs Indolent"
                )
```
```{r}
#Regression vs Progressive
EnhancedVolcano(toptable = res3,            
                x = "log2FoldChange",           
                y = "padj",                     
                lab = rownames(res3),
                pCutoff = 0.01,
                FCcutoff = 1,
                title = "Regression vs Progressive"
                )
```

Generating normalized counts

```{r}
dds <- estimateSizeFactors(dds)
```

```{r}
normalized_counts <- counts(dds, normalized=TRUE)
```

```{r}
dds.vst<-DESeq2::varianceStabilizingTransformation(dds)
expMatrix_Kwok <- SummarizedExperiment::assay(dds.vst)
expMatrix_Kwok <- as.data.frame(expMatrix_Kwok)
```

Heatmap of the sample-to-sample distances

```{r}
sampleDists <- dist(t(expMatrix_Kwok))
```
```{r}
sampleDistMatrix <- as.matrix(sampleDists)
```
```{r}
rownames(sampleDistMatrix) <- paste(metadata_Kwok$CLLtype)
colnames(sampleDistMatrix) <- paste(metadata_Kwok$CLLtype)
```
```{r}
library("RColorBrewer")
library("pheatmap")
```
```{r}
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         col=colors,
         main = "Heatmap of sample distances")
```

Heatmap of the most significant genes (p < 0.01)

```{r}
library("ComplexHeatmap")
library("RColorBrewer")
library("circlize")
library("grid")
```

```{r}
all_genes_results <- results(dds, pAdjustMethod = "BH", tidy = TRUE)
```
Extracting the table of differential genes
```{r}
DE_genes = all_genes_results %>% 
  as.data.frame() %>% 
  filter(padj < 0.01) %>% 
  arrange(desc(log2FoldChange), 
          desc(padj))
DE_genes <- DE_genes %>% remove_rownames %>% column_to_rownames(var="row")
```

```{r}
DE_genes <- merge(expMatrix_Kwok, DE_genes, by = "row.names")
```

```{r}
DE_genes <- DE_genes %>% remove_rownames %>% column_to_rownames(var="Row.names")
```

```{r}
DE_genes = DE_genes %>% 
  as.data.frame() %>% 
  filter(padj < 0.01) %>% 
  arrange(desc(log2FoldChange), 
          desc(padj))
```

```{r}
ann_col_info <- as.data.frame(metadata_Kwok)
anno_info_colors = list(
  CLLtype = c(Regression = "lightgrey", 
               Indolent = "black", Progressive = "red")
)
```

```{r}
DE_genes1 <- DE_genes
DE_genes1 <- DE_genes1[,-41:-46]

pheatmap(DE_genes1, 
         cluster_rows = FALSE,                       
         cluster_cols = TRUE, 
         show_rownames = FALSE, 
         show_colnames = FALSE,
         scale = "row",
         annotation_col = ann_col_info,
         annotation_colors = anno_info_colors,
         main = "Heatmap of different gene expression")
```

PCA analysis

```{r}
DE_genes1 <- t(DE_genes1)
DE_genes1 <- as.data.frame(DE_genes1)
DE_genes1 <- merge(DE_genes1, metadata_Kwok, by = "row.names")
DE_genes1 <- DE_genes1 %>% remove_rownames %>% column_to_rownames(var="Row.names")

```

```{r}
pca <- prcomp(DE_genes1[,c(1:836)])
summary(pca)
```

```{r}
autoplot(pca, data = DE_genes1, colour = "CLLtype", main = "PCA of CLL subtypes")
```
PCA 3d
```{r}
library(rgl)
library(pca3d)
```

```{r}
pca3d(pca, group = DE_genes2[,144] )
```

```{r}
pca3d(pca, group = DE_genes2[,144], 
    fancy= TRUE, bg= "black", 
    show.group.labels=TRUE,
    axes.color= "white", new= TRUE)
```


